### DISCUSSION TITLE
Nice push buttons with LEDs

### DISCUSSION THREAD

**Discussion by jlatex** [CrashOverride85/zc95#49] [Show and tell]
I wasn't a fan of the standard push buttons so went looking for something better. I found some nice looking buttons on Mouser,  E-Switch LP1OA1AB.

- The buttons require a 7.5mm cut out and the default on the front panel pcb is 7mm.  I used a circular file with a couple of turns and within seconds had enlarged the holes to 7.5mm
- The buttons are push fit rather than having any nuts, and they're really designed for a thicker panel, so they don't hold firmly in place. I secured them with a little dot of hotglue at the back
- The buttons are a little more mechanically bouncy, so sometimes they register two presses instead of one. I fixed this in software (and releases after v1.6 upstream have the fix for this).
- The buttons have an in-built LED. You can choose the colour. I chose Blue. They work by having two pins being the usual switch contacts, and two for the LED.  If you want them to light (not essential), then you need to run the LED positive to the 5v rail via LED for each resistor. The LED is rated at 20mA and therefore from 5v something like a 100R resistor for full brightness. In testing, this was far (far) too bright, and 1kR was a great compromise
- You could tie all the LED negatives to the 0v rail and the LEDs will be on all the time. 
- However my plan is to run them to the port expander on the ~~main~~ (edit:front panel) board which has spare pins and that way the LEDs will be under the software control - they could light up which ever switches have a function for example. I plan on doing a PR for that sometime
![20230813_125300](https://github.com/CrashOverride85/zc95/assets/89390295/00ecda32-44ac-49d2-9b49-945c21a6ccd8)

**Comment by CrashOverride85** [comment]
Those buttons are much nicer than the ones listed in the BoM, think I might get a set.

I like the idea of having the LEDs under software control, too. Although the only spare pins I can see are:
* 1 on the U8 of the main board. There's also P0-P2 on U8 which are only used for the accessory port, but I think it'd be a bit of headache firmware wise to make them switchable between the accessory port and the LEDs (assuming you don't want them on the acc port)
* 4 on U2 of the front panel board. Would mean the buttons would need to be connected to both main board and front panel

I'm wondering if either a second port expander, or a bigger one, could be put on the front panel board, and the buttons+LEDs wired to that. TBH I think it makes more sense for the buttons to be wired to that board instead of the main board anyway - not sure why I did it the way it is. As it would involve either a different or second port expander, it would have the benefit of being detectable by the firmware.

**Comment by jlatex** [comment]
ah, yes, I meant the one on the front panel creating a bit of a spaghetti mess of course.  (Perhaps with more expanders on the front panel it'd be worth having that one with more smt and use the assembly service)

**Comment by CrashOverride85** [comment]
Ahh cool, ok.
I'd been trying to avoid making a second board SMT to keep it easy to hand solder, and without the costs of the assembly service. But the ADC on that board is EOL soon (#36), and IIRC a lot of the options are much smaller SMT parts, so switching to SMT would make solving that problem easier too I guess.

**Comment by moarkaboom** [comment]
I was wondering if there would be a possibility to address the too bright LEDs by dimming them with a potentiometer or if thats not possible.
I have no electronics background at all, just some very basic knowledge, so this is the idea I came up with. Please excuse the crude sketch.
![image](https://github.com/user-attachments/assets/852d9fa7-486f-445e-a5d6-3a34b99d4c43)

**Reply by CrashOverride85** [reply]
The current SMT version of the front panel board (designed following this post) uses an LED driver IC to vary the brightness of the button LEDs based off a menu option on the box, solving the brightness issue that way.

For the original board, each LED needs its own resistor because the LEDs are individually controlled, so having a single resistor would mean the more LEDs that are on, the dimmer they would be (plus, 1 resistor per LED is a good idea anyway)
